File: src/webcore/view/viewcomponent.ts
"use strict";

import * as _ from "underscore";
import * as Handlebars from "handlebars";
import { IViewComponentOptions } from "./viewcomponentoptions";

export interface IView {
	[name: string]: any;
	_wasm_viewcomponent: any;
};

export class ViewComponent implements IView {
	public _wasm_context: any = null;
	public _wasm_viewcomponent: any = null;

	/**
	 * Конструктор представления.
	 * 
	 * @method constructor
	 * @param {any} context Контекст.
	 * @param {IViewOptions} options Свойства представления.
	 */
	constructor(context: any, options: IViewComponentOptions) {
		this._wasm_viewcomponent = function (self: any) {
			var viewComponentOptions = { tagname: "div", classname: "" };
			var ViewComponent = context.module.ViewComponent.extend("ViewComponent", {
				__construct: function(name, options) {
					this.__parent.__construct.call(this, name, options);
				},
				__destruct: function() {
					this.__parent.__destruct.call(this);
				},
				prerepaint: _.bind(self.prerepaint, self)
			});

			return new ViewComponent(options.id, _.defaults(options, viewComponentOptions));
		}(this) as any;

		this._wasm_context = context;
	}

	public initialize(): void {
		// Empty
	}

	public buildTemplate(html: string, context: any): void {
		this._wasm_viewcomponent.buildTemplate(_.template(Handlebars.compile(html)(context))());
	}

	public addRegion(name: string, options: any): void {
		this._wasm_viewcomponent.addRegion(name, options);
	}

	public getRegion(name: string): any {
		return this._wasm_viewcomponent.getRegion(name);
	}

	public getRegionNames(): any {
		return this._wasm_viewcomponent.getRegionNames();
	}

	public prerepaint(): void {
		// Empty
	}

	public getModel(): any {
		return this._wasm_viewcomponent.getModel();
	}

	public setModel(model: any): void {
		this._wasm_viewcomponent.setModel(model._wasm_model);
	}

	public addChild(view: IView): void {
		this._wasm_viewcomponent.addChild(view._wasm_viewcomponent);
	}

	public removeChild(view: IView): void {
		this._wasm_viewcomponent.removeChild(view._wasm_viewcomponent);
	}

	public getChildCount(): number {
		return this._wasm_viewcomponent.getChildCount();
	}

	public addChildView(view: IView): void {
		this._wasm_viewcomponent.addChildView(view._wasm_viewcomponent);
	}

	public removeChildView(view: IView): void {
		if (_.isNull(view._wasm_viewcomponent))
			return;

		this._wasm_viewcomponent.removeChildView(view._wasm_viewcomponent);
		delete view._wasm_viewcomponent;
	}

	public addEvent(targetId: string, type: any, callback: any): void {
		this._wasm_viewcomponent.addEvent(targetId, type, callback);
	}
};

File: src/webcore/view/viewcomponent.cpp
#include <sway/webcore/view/viewcomponent.h>

NAMESPACE_BEGIN(sway)
NAMESPACE_BEGIN(webcore)
NAMESPACE_BEGIN(view)

ViewComponent::ViewComponent(const std::string & name, const ViewComponentOptions & options)
	: TAbstractView<model::AbstractModel>(name)
	, _domElement(emscripten::val::null()) {

	_domElement = dom::Document::createElement(options.tagname);
	_domElement.setInnerHtml(options.htmlString);
	if (!options.id.empty()) _domElement.setAttribute("id", options.id);
	if (!options.classname.empty()) _domElement.addClassName(options.classname);
}

void ViewComponent::buildTemplate(const std::string & html) {
	_domElement.setInnerHtml(html);
}

void ViewComponent::addEvent(const std::string & targetId, EventTypes_t type, emscripten::val callback) {
	_handlers.push_back((struct EventHandler) {
		.targetId = targetId,
		.target = new EventTarget([](emscripten::val) { /* Empty */ }),
		.type = type
	});

	_handlers.back().target->setCallback(callback);
}

ViewComponent::~ViewComponent() {
	EM_ASM({console.log("DESTROY_VIEW " + UTF8ToString($0))}, getName().c_str());
	removeAll();
}

void ViewComponent::addRegion(const std::string & name, const RegionOptions & options) {
	_regions.insert(std::make_pair(name, std::make_shared<Region>(/*this, */options)));
}

RegionPtr_t ViewComponent::getRegion(const std::string & name) const {
	RegionIterator_t iter = _regions.find(name);
	if (iter != _regions.end())
		return iter->second;

	return nullptr;
}

RegionNameVec_t ViewComponent::getRegionNames() const {
	RegionNameVec_t keys;
	std::transform(_regions.begin(), _regions.end(), std::back_inserter(keys),
		[](const RegionMap_t::value_type & pair) {
			return pair.first;
		}
	);

	return keys;
}

void ViewComponent::addChildView(ViewComponent * child) {
	addChild(child);
	repaint();
}

void ViewComponent::removeChildView(ViewComponent * child) {
	removeChild(child);
	SAFE_DELETE(child);
}

void ViewComponent::update() {
	for (int i = 0; i < _handlers.size(); ++i) {
		EventHandler event = _handlers[i];
		event.target->addEventListener(event.targetId, stringize(event.type));
	}

	for (typename NodeMap_t::const_iterator iter = children.begin(); iter != children.end(); ++iter) {
		ViewComponent * child = static_cast<ViewComponent *>(iter->second);
		child->update();
	}
}

void ViewComponent::prerepaint() {
	// Default Implementation
}

void ViewComponent::repaint() {
	prerepaint();

	for (typename NodeMap_t::const_iterator iter = children.begin(); iter != children.end(); ++iter) {
		ViewComponent * child = static_cast<ViewComponent *>(iter->second);
		child->repaint();
	}
}

emscripten::val ViewComponent::getElement() {
	return _domElement;
}

NAMESPACE_END(view)
NAMESPACE_END(webcore)
NAMESPACE_END(sway)
